feat: Remove estimate gas

Thunkar opened the pull request:
The estimate gas call was polluting our intefaces, while the same results could be achieved by setting a combination of options in our simulation interface. This PR does just that. 

It also removes the forced fee estimation process when doing `prove`. When requesting a proof, the wallet should have already decided what `GasLimits` its going to use, which is a process that does not depend on the app at all. The ability to estimate fees during simulations requested by the app is kept so the information can be shown in a GUI, for example.

Thunkar commented on yarn-project/test-wallet/src/wallet/test_wallet.ts:
```diff
@@ -161,12 +161,17 @@ export abstract class BaseTestWallet extends BaseWallet {
     executionPayload: ExecutionPayload,
     opts: SimulateMethodOptions,
   ): Promise<TxSimulationResult> {
+    if (this.simulatedSimulations && opts.fee?.estimateGas) {
```
This is not necessarily true, but more data is needed (specially when this feature comes online in public). ATM they're never used simultaneously, but I thought for the `TestWallet` this was a reasonable protection.

benesjan replied:
Would document the simulatedSimulations variable.

I don't think a code reader that didn't participate in the original simulated simulations discussions would be able to figure this out.

Thunkar replied:
I will, but this is 100% internal and not the form it's going to take even on the TestWallet

benesjan commented on yarn-project/aztec.js/src/contract/interaction_options.ts:
```diff
@@ -50,3 +51,30 @@ export type ProfileMethodOptions = SimulateMethodOptions & {
+/**
+ * Represents the result type of a simulation.
+ * By default, it will just be the return value of the simulated function
+ * so contract interfaces behave as plain functions. If `includeMetadata` is set to true in `SimulateMethodOptions` on the input of `simulate(...)`,
```
Find the "so contract interfaces behave as plain functions" hard to grok.

benesjan commented on yarn-project/cli-wallet/src/utils/wallet.ts:
```diff
@@ -50,11 +49,19 @@ export class CLIWallet extends BaseWallet {
     from: AztecAddress,
     userFee?: UserFeeOptions,
-  ): Promise<TxExecutionRequest> {
+  ): Promise<{ txRequest: TxExecutionRequest; gasEstimationsSimulationResults?: TxSimulationResult }> {
```
This function name is now broken because it sneaks in the return value the gas estimation tingy. This is not clear from the function name.

benesjan commented on boxes/boxes/vanilla/app/embedded-wallet.ts:
```diff
@@ -271,12 +271,15 @@ export class EmbeddedWallet extends BaseWallet {
       artifact,
     } = await this.getFakeAccountDataFor(opts.from);
-    const fee = await this.getFeeOptions(
+    const { fee, gasEstimationsSimulationResults } = await this.getFeeOptions(
```
The fetFeeOptions name is misleading as it returns what I would not expect - gasEstimationsSimulationResults

Thunkar commented on boxes/boxes/vanilla/app/embedded-wallet.ts:
```diff
@@ -58,6 +64,26 @@ export class EmbeddedWallet extends BaseWallet {
     return account;
   }
 
+  override async getDefaultFeeOptions(
```
This makes a lot more sense: the embedded wallet wants to replace the fee payment method, so it does so in a single place

benesjan replied:
I would document here what does default even mean, what happens with the fee arg on the input and why is the address ignored.

I looked around and the function on BaseWallet also don't explain what default means here:

<img width="981" height="355" alt="image" src="https://github.com/user-attachments/assets/0265abfa-b1e7-486d-8ee0-14a1ba28c59b" />

Thunkar commented on yarn-project/pxe/src/contract_function_simulator/contract_function_simulator.ts:
```diff
@@ -449,7 +464,7 @@ export async function generateSimulatedProvingResult(
   const publicInputs = new PrivateKernelTailCircuitPublicInputs(
     constantData,
-    /*gasUsed=*/ new Gas(0, 0),
+    /*gasUsed=*/ gasUsed,
```
When doing kernelless simulations, used gas used to be 0...since we weren't doing the kernel's job properly. This is now fixed

Thunkar commented on yarn-project/cli-wallet/src/cmds/deploy.ts:
```diff
@@ -64,59 +64,60 @@ export async function deploy(
     skipInstancePublication,
   };
 
-  if (feeOpts.estimateOnly) {
```
Bringing this in line with the rest of the commands: Simulation is unconditionally performed to avoid sending invalid txs

benesjan commented on boxes/boxes/vanilla/app/embedded-wallet.ts:
```diff
@@ -58,6 +64,26 @@ export class EmbeddedWallet extends BaseWallet {
+    const maxFeesPerGas =
+      fee?.gasSettings?.maxFeesPerGas ??
+      (await this.aztecNode.getCurrentBaseFees()).mul(1 + this.baseFeePadding);
+    const sponsoredPFCContract =
```
```suggestion
    const sponsoredFPCContract =
```

benesjan commented on playground/src/wallet/embedded_wallet.ts:
```diff
@@ -232,7 +232,9 @@ export class EmbeddedWallet extends BaseWallet {
     const executionOptions = { txNonce: Fr.random(), cancellable: false };
     const { account: fromAccount, instance, artifact } = await this.getFakeAccountDataFor(opts.from);
-    const fee = await this.getFeeOptions(fromAccount, executionPayload, opts.fee, executionOptions);
+    const fee = opts.fee?.estimateGas
```
I would expect the name "fee" to be an integer. Would rename this to feeOptions everywhere.

Thunkar replied:
Changed

benesjan commented on boxes/boxes/vanilla/app/embedded-wallet.ts:
```diff
@@ -58,6 +64,26 @@ export class EmbeddedWallet extends BaseWallet {
 
+  override async getDefaultFeeOptions(
+    _address: AztecAddress,
+    fee: UserFeeOptions | SimulationUserFeeOptions | undefined
```
Would document why is fee of different types. Like why use the type with and without "simulation". Is simulation the "fake" run used to just estimate gas?

Thunkar replied:
This type duplicity ended up being unnecessary

benesjan commented on yarn-project/aztec.js/src/contract/contract_function_interaction.ts:
```diff
@@ -105,6 +81,10 @@ export class ContractFunctionInteraction extends BaseContractInteraction {
    * @param options - An optional object containing additional configuration for the transaction.
    * @returns The result of the transaction as returned by the contract function.
    */
```
Could you document here why the overload?

Also have you considered splitting this function into multiple simpler ones? It's very hard to follow the types here.

This also seems like a bit of a lie given that you modify it in this function if the flags in options are set:

"The result of the transaction as returned by the contract function."

Thunkar replied:
Documented. I think this is somewhat jarring if you see the implementation, but in actual use is intuitive and clear (also the type system will help you)

benesjan commented on yarn-project/aztec.js/src/contract/contract_function_interaction.ts:
```diff
@@ -105,6 +81,10 @@ export class ContractFunctionInteraction extends BaseContractInteraction {
    * @returns The result of the transaction as returned by the contract function.
    */
+  public async simulate<T extends SimulateMethodOptions>(
+    options: T,
```
I think these options make the code hard to navigate as it's tricky to figure out what code branch is triggered when you stumble upon it in the code (you have to hunt for where this is actually set). I know it's all over the place in `Aztec.js` but it just seems like a bad software design choice.

Thunkar replied:
Documented, hope it's clearer now. I think it's useful and not a bad choice at all.

benesjan commented on yarn-project/aztec.js/src/wallet/base_wallet.ts:
```diff
@@ -188,21 +141,15 @@ export abstract class BaseWallet implements Wallet {
-    /**
-     * Determes if the provided instance data is already a contract instance with an address.
-     */
+    /** Determes if the provided instance data is already a contract instance with an address. */
```
Determes? Didn't the spellcheck extension detect this?

Thunkar replied:
🤷

benesjan commented on yarn-project/aztec.js/src/wallet/base_wallet.ts:
```diff
@@ -78,12 +79,11 @@ export abstract class BaseWallet implements Wallet {
   protected async createTxExecutionRequestFromPayloadAndFee(
```
Again would expect "fee" to be an integer based on the name.

Thunkar replied:
Changed

benesjan commented on yarn-project/aztec.js/src/wallet/base_wallet.ts:
```diff
@@ -94,90 +94,43 @@ export abstract class BaseWallet implements Wallet {
+   * Returns unreasonably high gas limits in order to execute a simulation
+   * with the goal of estimating its gas cost
+   * @param address - The address of the account requesting the fee options
    * @param fee - User-provided fee options.
```
This should document what happens if the fee arg is undefined.

Thunkar replied:
I've tried to clarify and improve the wording in the different wallets

Thunkar replied:
The thing with `getDefaultFeeOptions` is that it does exactly what it says on the tin. What `default` means can be totally different for different wallets, but essentially tries to fill in the gaps the user has left when they provide incomplete (or completely missing) options for fee payment

benesjan commented on yarn-project/aztec.js/src/contract/deploy_method.ts:
```diff
@@ -248,15 +265,24 @@ export class DeployMethod<TContract extends ContractBase = Contract> extends Bas
-   * @returns Gas limits.
+   * Simulate the deployment (used for gas estimations)
+   * @param options - An optional object containing additional configuration for the transaction.
+   * @returns The result of the transaction as returned by the contract function.
```
"The result of the transaction as returned by the contract function."

This sounds strange. Is contract function returning transaction result?

Also here the function would be a constructor so think it's useful to make this more specific.

Thunkar replied:
Clarified

benesjan reviewed (commented):
Sorry if I am being too difficult here but I am not sure the code is clear enough to follow

benesjan commented on yarn-project/aztec.js/src/contract/deploy_method.ts:
```diff
@@ -45,6 +46,22 @@ export type DeployOptions = {
+  skipTxValidation?: boolean;
+  /** Whether to ensure the fee payer is not empty and has enough balance to pay for the fee. */
+  skipFeeEnforcement?: boolean;
+  /** Whether to include metadata such as offchain effects and performance statistics (e.g. timing information of the different circuits and oracles) in
```
Just a detail but would stick to 120 chars per line as then generally people need to scroll sideways on screens.

benesjan commented on yarn-project/aztec.js/src/contract/deploy_method.ts:
```diff
@@ -248,15 +265,26 @@ export class DeployMethod<TContract extends ContractBase = Contract> extends Bas
+    return {
+      stats: simulatedTx.stats!,
+      offchainEffects: collectOffchainEffects(simulatedTx.privateExecutionResult),
+      result: undefined,
```
```suggestion
      result: undefined, // Constructors never return any value
```

Would this ^ be correct? The `result` here is a return value, right?

benesjan commented on yarn-project/aztec.js/src/contract/interaction_options.ts:
```diff
@@ -29,7 +30,9 @@ export type SendMethodOptions = RequestMethodOptions & {
  */
-export type SimulateMethodOptions = Pick<SendMethodOptions, 'from' | 'authWitnesses' | 'capsules' | 'fee'> & {
+export type SimulateMethodOptions = Omit<SendMethodOptions, 'fee'> & {
+  /** The fee options for the transaction. */
```
These options are just for simulation purposes so having there "for the transaction" is incorrect, no?

Thunkar replied:
You always simulate a tx, not a function (at least account entrypoint+function). UtilitySimulationResult is for single functions

benesjan commented on yarn-project/aztec.js/src/contract/interaction_options.ts:
```diff
@@ -29,7 +30,9 @@ export type SendMethodOptions = RequestMethodOptions & {
-export type SimulateMethodOptions = Pick<SendMethodOptions, 'from' | 'authWitnesses' | 'capsules' | 'fee'> & {
+export type SimulateMethodOptions = Omit<SendMethodOptions, 'fee'> & {
+  /** The fee options for the transaction. */
+  fee?: SimulationUserFeeOptions;
```
Would also rename to feeOptions to have the naming consistent.

Thunkar replied:
Its inside an options object, I profoundly dislike `options.feeOptions`

benesjan commented on yarn-project/aztec.js/src/wallet/base_wallet.ts:
```diff
@@ -94,90 +94,50 @@ export abstract class BaseWallet implements Wallet {
-   * @param fee - User-provided fee options.
-   */
-  private async getDefaultFeeOptions(account: Account, fee: UserFeeOptions | undefined): Promise<FeeOptions> {
+  protected async getDefaultFeeOptions(
```
I would rename this function to `populateMissingFeeOptionsWithDefaults` or just `populateMissingFeeOptions`. It's verbose but that's what it does. If a user provides partially filled user fee options on the input then calling the return value a "default fee options" is a lie because some of it have been user-modified.

Thunkar replied:
This is being reworked on the next PR. The wallet will just receive some hints and it will have to construct the actual `FeeOptions` here based on serializable information

benesjan commented on yarn-project/aztec.js/src/wallet/base_wallet.ts:
```diff
@@ -94,90 +94,50 @@ export abstract class BaseWallet implements Wallet {
-   * @param executionPayload - The execution payload to calculate the gas for
-   * @param opts - Options.
-   * @returns Gas limits.
+   * Returns default values for the transaction fee options
+   * if they were omitted by the user.
+   * @param address - The account building the transaction, usually the one paying the fee
+   * @param userFeeOptions - User-provided fee options, which might be incomplete
```
```suggestion
   * User-provided fee options provided on the input contain the same values as the return value type with
   * the difference that the values in UserFeeOptions are optional. This function populates the missing values
   * in UserFeeOptions with defaults.
   *
   * @param address - The account building the transaction, usually the one paying the fee
   * @param userFeeOptions - User-provided fee options to use as a basis for the output
```
Before I needed to go check what exactly are UserFeeOptions to understand what this does. Think with this change it's now unnecessary.

Please copy this to other implementations of this func if you agree with it.

benesjan replied:
Maybe we could even rename `UserFeeOptions` to something like`PartialFeeOptions`. Then just the types of inputs and outputs of the function would make it quite clear what it does. WDYT?

Thunkar replied:
This type is being completely reworked into `FeeOptionsHint` in the next PR

benesjan commented on boxes/boxes/vanilla/app/embedded-wallet.ts:
```diff
@@ -58,6 +64,36 @@ export class EmbeddedWallet extends BaseWallet {
+   * This wallet will use the sponsoredFPC payment method
+   * unless otherwise stated, which is why the address parameter
+   * (who is paying the fee) is unused
+   * @param _address - Unused
```
Would it be correct to rename this arg as `_feePayer`?

`_address` is too vague.

Thunkar replied:
Not really. In the case of an fpc, feePayer is the contract itself

benesjan commented on yarn-project/aztec.js/src/wallet/base_wallet.ts:
```diff
@@ -94,90 +94,50 @@ export abstract class BaseWallet implements Wallet {
+   * with the goal of estimating its gas cost. It will otherwise try to respect
+   * the user-specified fee options, filling the gaps with default values as needed.
+   * @param address - The address of the account requesting the fee options
+   * @param userFeeOptions - User-provided fee options.
```
```suggestion
   * @param userFeeOptions - User-provided fee options to use as a basis for the fully populated `FeeOptions` type.
```

benesjan commented on yarn-project/cli-wallet/src/cmds/create_account.ts:
```diff
@@ -81,10 +81,16 @@ export async function createAccount(
+      contractAddressSalt: salt,
     };
+    //  Estimated gas usage:    da=14081,l2=8634,teardownDA=0,teardownL2=0
+    // Maximum total tx fee:   13210020
```
Is this a forgotten comment here?

Thunkar replied:
Yes

benesjan commented on yarn-project/cli-wallet/src/cmds/create_account.ts:
```diff
@@ -81,10 +81,16 @@ export async function createAccount(
       from,
       fee: userFeeOptions,
+      // Do not mix the deployer in the address, since the account
+      // was created (and thus its address was fixed) like this
```
Given that the account was already created on line 29 why are we recreating the DeployOptions here? Is it just because we want to print them out?

Thunkar replied:
Creating an account just registers it in the wallet. You have to provide then correct deployOptions so that the contract on chain can verify correct address derivation. We usually do it in a single step, but the CLI wallet supports creating accounts and never deploying them (fully private)

benesjan replied:
That would be a great comment in the code 😊

benesjan commented on yarn-project/cli-wallet/src/cmds/create_account.ts:
```diff
@@ -63,7 +63,7 @@ export async function createAccount(
   let tx;
   let txReceipt;
   if (!registerOnly) {
```
What does this flag mean?

I am not able to understand this func without docs.

Thunkar replied:
<img width="708" height="72" alt="Screenshot 2025-09-23 at 12 20 21" src="https://github.com/user-attachments/assets/399c54a9-8bc4-4d76-bb5a-4a995c5cf721" />

Sometimes, reading the code pays off 😛

benesjan commented on yarn-project/end-to-end/src/e2e_fees/gas_estimation.test.ts:
```diff
@@ -51,13 +52,14 @@ describe('e2e_fees gas_estimation', () => {
 
-  // Sends two tx with transfers of public tokens: one with estimateGas on, one with estimateGas off
-  const sendTransfers = (paymentMethod: FeePaymentMethod) =>
+  // Sends two tx with transfers of public tokens: one with limits based on the estimate, another one without
```
```suggestion
  // Sends two txs with transfers of public tokens: one with limits based on the estimate, another one without
```

benesjan commented on yarn-project/pxe/src/pxe_service/pxe_service.ts:
```diff
@@ -620,24 +620,19 @@ export class PXEService implements PXE {
-    privateExecutionResult?: PrivateExecutionResult,
-  ): Promise<TxProvingResult> {
+  public proveTx(txRequest: TxExecutionRequest): Promise<TxProvingResult> {
+    let privateExecutionResult: PrivateExecutionResult;
```
I think declaring the var here instead of on line 634 is just a mistake.

Thunkar replied:
No, since it's used in the catch block (the process can fail at any stage, and all the extra info is good)

Thunkar committed:
feat: Remove estimate gas

The estimate gas call was polluting our intefaces, while the same results could be achieved by setting a combination of options in our simulation interface. This PR does just that.

It also removes the forced fee estimation process when doing `prove`. When requesting a proof, the wallet should have already decided what `GasLimits` its going to use, which is a process that does not depend on the app at all. The ability to estimate fees during simulations requested by the app is kept so the information can be shown in a GUI, for example.

Co-authored-by: Gregorio Juliana <gregojquiros@gmail.com>
Co-authored-by: thunkar <gregojquiros@gmail.com>
